src/tests/nativeaot/SmokeTests/PInvoke/PInvoke.cs

```diff
@@ -1053,6 +1104,72 @@ override protected bool ReleaseHandle()
     [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
     public interface IComInterface
     {
+        void DoWork();
+    }
+
+    public class ComObject : IComInterface
+    {
+        public void DoWork()
+        {
+            Console.WriteLine("Do work called.");
+        }
+    }
+
+    internal unsafe class SimpleComWrapper : ComWrappers
+    {
+        static ComInterfaceEntry* wrapperEntry;
+
+        static SimpleComWrapper()
+        {
+            IntPtr* vtbl = (IntPtr*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(IComInterfaceVtbl), sizeof(IComInterfaceVtbl));
+            GetIUnknownImpl(out vtbl[0], out vtbl[1], out vtbl[2]);
+            vtbl[3] = (IntPtr)(delegate* unmanaged<IntPtr, void>)&IComInterfaceProxy.DoWork;
+
+            var comInterfaceEntryMemory = RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(IComInterfaceVtbl), sizeof(ComInterfaceEntry));
```

Comment:
```suggestion
            IntPtr* vtbl = (IntPtr*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(IComInterface), 4 * sizeof(IntPtr));
            GetIUnknownImpl(out vtbl[0], out vtbl[1], out vtbl[2]);
            vtbl[3] = (IntPtr)(delegate* unmanaged<IntPtr, void>)&IComInterfaceProxy.DoWork;

            var comInterfaceEntryMemory = RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(IComInterface), sizeof(ComInterfaceEntry));
```
and also address https://github.com/dotnet/runtimelab/pull/653/files#r593240496

The Vtbl structs are unnecessary overhead. The recomendation is to just allocate IntPtr blocks. It is what CsWinRT does today - they have started with the Vtbl structs and found that there is a lot of overhead that comes with them.

Reply:
I disagree in general, since I think this kind of structures should be generated and manually created, that way no problem would be with overhead IMO. And I personally think that this is too low-level, but maybe I'm missing some interesting performance characteristics.

Reply:
The Vtbl structures have statics footprint overhead. They increase size for AOT and startup time for JIT.

https://github.com/microsoft/CsWinRT/pull/492 has some numbers how much it saved in CsWinRT to get rid of them.